Container startup failed for *CdcNetworkIntegrationTest

_master_ (commit 0fbe205b49c75a4f1df05b2f0d4ad29f81c43f04)

`*CdcNetworkIntegrationTest`s intermittently fail with `Container startup failed`. I see following tests to fail with that:
- `com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail]`
- `com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringSnapshotting_then_jetSourceIsStuckUntilReconnect[reconnect]`
- `com.hazelcast.jet.cdc.postgres.PostgresCdcNetworkIntegrationTest.when_databaseShutdownOrLongDisconnectDuringBinlogReading[fail]`
- `com.hazelcast.jet.cdc.postgres.PostgresCdcNetworkIntegrationTest.when_shortNetworkDisconnectDuringSnapshotting_then_connectorDoesNotNoticeAnything[reconnect]`
- `com.hazelcast.jet.cdc.postgres.PostgresCdcNetworkIntegrationTest.when_shortNetworkDisconnectDuringSnapshotting_then_connectorDoesNotNoticeAnything[reconnect w/ state reset]`

Failed on nightly run, for example: http://jenkins.hazelcast.com/job/jet-oss-master-nightly/274/testReport/junit/com.hazelcast.jet.cdc.mysql/MySqlCdcNetworkIntegrationTest/when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally_fail_/

Stacktrace (for `com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail]`):
```
org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:330)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311)
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.initToxiproxy(MySqlCdcNetworkIntegrationTest.java:369)
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally(MySqlCdcNetworkIntegrationTest.java:237)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.lang.NullPointerException
		at java.util.concurrent.ConcurrentHashMap.replaceNode(ConcurrentHashMap.java:1106)
		at java.util.concurrent.ConcurrentHashMap.remove(ConcurrentHashMap.java:1097)
		at java.util.Collections$SetFromMap.remove(Collections.java:5514)
		at org.testcontainers.utility.ResourceReaper.removeNetwork(ResourceReaper.java:359)
		at org.testcontainers.utility.ResourceReaper.removeNetworkById(ResourceReaper.java:322)
		at org.testcontainers.containers.Network$NetworkImpl.close(Network.java:102)
		at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally(MySqlCdcNetworkIntegrationTest.java:268)
		... 12 more
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323)
	... 15 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:498)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 16 more
Caused by: com.github.dockerjava.api.exception.NotFoundException: {"message":"could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:287)
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:271)
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.post(OkHttpInvocationBuilder.java:129)
	at com.github.dockerjava.core.exec.CreateNetworkCmdExec.execute(CreateNetworkCmdExec.java:27)
	at com.github.dockerjava.core.exec.CreateNetworkCmdExec.execute(CreateNetworkCmdExec.java:12)
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.lambda$wrappedCommand$14(AuditLoggingDockerClient.java:102)
	at com.sun.proxy.$Proxy53.exec(Unknown Source)
	at org.testcontainers.containers.Network$NetworkImpl.create(Network.java:91)
	at org.testcontainers.containers.Network$NetworkImpl.getId(Network.java:62)
	at org.testcontainers.containers.GenericContainer.applyConfiguration(GenericContainer.java:811)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:359)
	... 18 more
```

Standard output (for `com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail]`):
```
Finished Running Test: when_noDatabaseToConnectTo[fail] in 12.814 seconds.
Started Running Test: when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail]
2020-10-08 23:53:36,324 [ INFO] [when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally] [d.1.0]]: Creating container for image: shopify/toxiproxy:2.1.0
2020-10-08 23:53:36,326 [ERROR] [when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally] [d.1.0]]: Could not start container
com.github.dockerjava.api.exception.NotFoundException: {"message":"could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network"}

	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:287) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:271) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.post(OkHttpInvocationBuilder.java:129) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.CreateNetworkCmdExec.execute(CreateNetworkCmdExec.java:27) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.CreateNetworkCmdExec.execute(CreateNetworkCmdExec.java:12) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35) ~[testcontainers-1.14.3.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_251]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_251]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_251]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_251]
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.lambda$wrappedCommand$14(AuditLoggingDockerClient.java:102) ~[testcontainers-1.14.3.jar:?]
	at com.sun.proxy.$Proxy53.exec(Unknown Source) ~[?:?]
	at org.testcontainers.containers.Network$NetworkImpl.create(Network.java:91) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.Network$NetworkImpl.getId(Network.java:62) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.applyConfiguration(GenericContainer.java:811) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:359) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325) ~[testcontainers-1.14.3.jar:?]
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81) [duct-tape-1.0.8.jar:?]
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323) [testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311) [testcontainers-1.14.3.jar:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.initToxiproxy(MySqlCdcNetworkIntegrationTest.java:369) [test-classes/:?]
	at com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest.when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally(MySqlCdcNetworkIntegrationTest.java:237) [test-classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_251]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_251]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_251]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_251]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114) [hazelcast-4.0.3-tests.jar:4.0.3]
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106) [hazelcast-4.0.3-tests.jar:4.0.3]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_251]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
BuildInfo right after when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail](com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest): BuildInfo{version='4.0.3', build='20200921', buildNumber=20200921, revision=59ae831, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='4.4-SNAPSHOT', build='20201008', revision='0fbe205'}}
Hiccups measured while running test 'when_networkDisconnectDuringBinlogRead_then_connectorReconnectsInternally[fail](com.hazelcast.jet.cdc.mysql.MySqlCdcNetworkIntegrationTest):'
23:53:35, accumulated pauses: 61 ms, max pause: 13 ms, pauses over 1000 ms: 0
23:53:40, accumulated pauses: 39 ms, max pause: 2 ms, pauses over 1000 ms: 0
23:53:45, accumulated pauses: 96 ms, max pause: 73 ms, pauses over 1000 ms: 0
```

--- a/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/AbstractCdcIntegrationTest.java
+++ b/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/AbstractCdcIntegrationTest.java
@@ -18,6 +18,7 @@ package com.hazelcast.jet.cdc;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.InspectContainerResponse;
 import com.hazelcast.function.SupplierEx;
 import com.hazelcast.jet.core.JetTestSupport;
 import com.hazelcast.jet.core.Processor;
@@ -31,14 +32,15 @@ import org.junit.experimental.categories.Category;
 import org.junit.rules.TestName;
 import org.testcontainers.DockerClientFactory;
 import org.testcontainers.containers.GenericContainer;
-import org.testcontainers.utility.ResourceReaper;
 
 import javax.annotation.Nonnull;
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
@@ -125,21 +127,68 @@ public class AbstractCdcIntegrationTest extends JetTestSupport {
         }
     }
 
+    protected static int fixPortBinding(GenericContainer<?> container, Integer defaultPort) {
+        int port = container.getMappedPort(defaultPort);
+        container.setPortBindings(Collections.singletonList(port + ":" + defaultPort));
+        return port;
+    }
+
+    protected static void startContainer(GenericContainer<?> container) {
+        container.start();
+        waitForContainerToStart(container);
+    }
+
     /**
      * Stops a container via the `docker stop` command. Necessary because the
      * {@code stop()} method of test containers is implemented via `docker kill`
      * and this matters for some tests.
      */
     protected static void stopContainer(GenericContainer<?> container) {
-        String containerId = container.getContainerId();
         DockerClient dockerClient = DockerClientFactory.instance().client();
-        dockerClient.stopContainerCmd(containerId)
-                .exec();
-        dockerClient.removeContainerCmd(containerId)
-                .withRemoveVolumes(true)
-                .withForce(true)
-                .exec();
-        ResourceReaper.instance().unregisterContainer(containerId);
+        dockerClient.stopContainerCmd(container.getContainerId()).exec();
+
+        waitForContainerToStop(container);
+
+        container.stop(); //allow the resource reaper to clean its registries
+    }
+
+    private static void waitForContainerToStart(GenericContainer<?> container) {
+        String containerId = container.getContainerId();
+
+        DockerClient dockerClient = DockerClientFactory.instance().client();
+        boolean running;
+        do {
+            InspectContainerResponse containerInfo = dockerClient.inspectContainerCmd(containerId).exec();
+            running = containerInfo.getState() != null && Boolean.TRUE.equals(containerInfo.getState().getRunning());
+            if (!running) {
+                System.out.println("Waiting for container " + containerId + " to start ...");
+                sleepABit();
+            }
+        } while (!running);
+    }
+
+    private static void waitForContainerToStop(GenericContainer<?> container) {
+        String containerId = container.getContainerId();
+
+        DockerClient dockerClient = DockerClientFactory.instance().client();
+        boolean running;
+        do {
+            InspectContainerResponse containerInfo = dockerClient.inspectContainerCmd(containerId).exec();
+            running = containerInfo.getState() != null && Boolean.TRUE.equals(containerInfo.getState().getRunning());
+            if (running) {
+                System.out.println("Waiting for container " + containerId + " to stop ...");
+                sleepABit();
+            }
+        } while (running);
+    }
+
+    private static void sleepABit() {
+        try {
+            TimeUnit.MILLISECONDS.sleep(250);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw sneakyThrow(e);
+        }
     }
 
     protected <T> T namedTestContainer(GenericContainer<?> container) {

--- a/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/AbstractCdcIntegrationTest.java
+++ b/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/AbstractCdcIntegrationTest.java
@@ -18,7 +18,6 @@ package com.hazelcast.jet.cdc;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.github.dockerjava.api.DockerClient;
-import com.github.dockerjava.api.command.InspectContainerResponse;
 import com.hazelcast.function.SupplierEx;
 import com.hazelcast.jet.core.JetTestSupport;
 import com.hazelcast.jet.core.Processor;
@@ -40,7 +39,6 @@ import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
-import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
@@ -133,11 +131,6 @@ public class AbstractCdcIntegrationTest extends JetTestSupport {
         return port;
     }
 
-    protected static void startContainer(GenericContainer<?> container) {
-        container.start();
-        waitForContainerToStart(container);
-    }
-
     /**
      * Stops a container via the `docker stop` command. Necessary because the
      * {@code stop()} method of test containers is implemented via `docker kill`
@@ -147,48 +140,7 @@ public class AbstractCdcIntegrationTest extends JetTestSupport {
         DockerClient dockerClient = DockerClientFactory.instance().client();
         dockerClient.stopContainerCmd(container.getContainerId()).exec();
 
-        waitForContainerToStop(container);
-
         container.stop(); //allow the resource reaper to clean its registries
-    }
-
-    private static void waitForContainerToStart(GenericContainer<?> container) {
-        String containerId = container.getContainerId();
-
-        DockerClient dockerClient = DockerClientFactory.instance().client();
-        boolean running;
-        do {
-            InspectContainerResponse containerInfo = dockerClient.inspectContainerCmd(containerId).exec();
-            running = containerInfo.getState() != null && Boolean.TRUE.equals(containerInfo.getState().getRunning());
-            if (!running) {
-                System.out.println("Waiting for container " + containerId + " to start ...");
-                sleepABit();
-            }
-        } while (!running);
-    }
-
-    private static void waitForContainerToStop(GenericContainer<?> container) {
-        String containerId = container.getContainerId();
-
-        DockerClient dockerClient = DockerClientFactory.instance().client();
-        boolean running;
-        do {
-            InspectContainerResponse containerInfo = dockerClient.inspectContainerCmd(containerId).exec();
-            running = containerInfo.getState() != null && Boolean.TRUE.equals(containerInfo.getState().getRunning());
-            if (running) {
-                System.out.println("Waiting for container " + containerId + " to stop ...");
-                sleepABit();
-            }
-        } while (running);
-    }
-
-    private static void sleepABit() {
-        try {
-            TimeUnit.MILLISECONDS.sleep(250);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw sneakyThrow(e);
-        }
     }
 
     protected <T> T namedTestContainer(GenericContainer<?> container) {

--- a/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
@@ -136,7 +136,7 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
             assertTrue(jet.getMap("results").isEmpty());
 
             // and DB starts
-            startContainer(mysql);
+            mysql.start();
             try {
                 // then source connects successfully
                 assertEqualsEventually(() -> jet.getMap("results").size(), 4);
@@ -347,7 +347,7 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
         if (network != null) {
             mysql = mysql.withNetwork(network);
         }
-        startContainer(mysql);
+        mysql.start();
         return mysql;
     }
 

--- a/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcNetworkIntegrationTest.java
@@ -47,22 +47,16 @@ import org.testcontainers.containers.MySQLContainer;
 import org.testcontainers.containers.Network;
 import org.testcontainers.containers.ToxiproxyContainer;
 
-import java.io.IOException;
-import java.net.ServerSocket;
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeoutException;
 import java.util.function.Consumer;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 
 import static com.hazelcast.jet.Util.entry;
 import static com.hazelcast.jet.core.JobStatus.RUNNING;
@@ -114,18 +108,16 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
     @After
     public void after() {
         if (mysql != null) {
-            stopContainer(mysql);
+            mysql.stop();
         }
     }
 
     @Test
     public void when_noDatabaseToConnectTo() throws Exception {
-        mysql = initMySql(null, 0);
+        mysql = initMySql(null, null);
+        int port = fixPortBinding(mysql, MYSQL_PORT);
         String containerIpAddress = mysql.getContainerIpAddress();
         stopContainer(mysql);
-        mysql = null;
-
-        int port = findRandomOpenPortInRange(MYSQL_PORT + 100, MYSQL_PORT + 1000);
 
         Pipeline pipeline = initPipeline(containerIpAddress, port);
 
@@ -144,7 +136,7 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
             assertTrue(jet.getMap("results").isEmpty());
 
             // and DB starts
-            mysql = initMySql(null, port);
+            startContainer(mysql);
             try {
                 // then source connects successfully
                 assertEqualsEventually(() -> jet.getMap("results").size(), 4);
@@ -193,8 +185,9 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
     @Test
     public void when_databaseShutdownDuringSnapshotting() throws Exception {
-        int port = findRandomOpenPort();
-        mysql = initMySql(null, port);
+        mysql = initMySql(null, null);
+        int port = fixPortBinding(mysql, MYSQL_PORT);
+
         Pipeline pipeline = initPipeline(mysql.getContainerIpAddress(), port);
         // when job starts
         JetInstance jet = createJetMembers(2)[0];
@@ -207,7 +200,6 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
         // and DB is stopped
         stopContainer(mysql);
-        mysql = null;
 
         boolean neverReconnect = reconnectBehavior.getMaxAttempts() == 0;
         if (neverReconnect) {
@@ -267,8 +259,9 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
     @Test
     public void when_databaseShutdownDuringBinlogReading() throws Exception {
-        int port = findRandomOpenPort();
-        mysql = initMySql(null, port);
+        mysql = initMySql(null, null);
+        int port = fixPortBinding(mysql, MYSQL_PORT);
+
         Pipeline pipeline = initPipeline(mysql.getContainerIpAddress(), port);
         // when connector is up and transitions to binlog reading
         JetInstance jet = createJetMembers(2)[0];
@@ -280,7 +273,6 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
 
         // and DB is stopped
         stopContainer(mysql);
-        mysql = null;
 
         boolean neverReconnect = reconnectBehavior.getMaxAttempts() == 0;
         if (neverReconnect) {
@@ -355,7 +347,7 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
         if (network != null) {
             mysql = mysql.withNetwork(network);
         }
-        mysql.start();
+        startContainer(mysql);
         return mysql;
     }
 
@@ -385,31 +377,6 @@ public class MySqlCdcNetworkIntegrationTest extends AbstractCdcIntegrationTest {
             statement.executeBatch();
             connection.commit();
         }
-    }
-
-    private static int findRandomOpenPort() throws Exception {
-        try (ServerSocket socket = new ServerSocket(0)) {
-            return socket.getLocalPort();
-        }
-    }
-
-    private static int findRandomOpenPortInRange(int fromInclusive, int toExclusive) throws IOException {
-        List<Integer> randomizedPortsInRange = IntStream.range(fromInclusive, toExclusive)
-                .boxed()
-                .collect(Collectors.toList());
-        Collections.shuffle(randomizedPortsInRange);
-
-        for (int port : randomizedPortsInRange) {
-            try {
-                ServerSocket serverSocket = new ServerSocket(port);
-                serverSocket.close();
-                return port;
-            } catch (Exception e) {
-                //swallow, expected
-            }
-        }
-
-        throw new IOException("No free port in range [" + fromInclusive + ", " + toExclusive + ")");
     }
 
     @SuppressWarnings("StatementWithEmptyBody")

--- a/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
@@ -48,20 +48,14 @@ import org.testcontainers.containers.PostgreSQLContainer;
 import org.testcontainers.containers.ToxiproxyContainer;
 
 import javax.annotation.Nonnull;
-import java.io.IOException;
-import java.net.ServerSocket;
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.function.Consumer;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 
 import static com.hazelcast.jet.Util.entry;
 import static com.hazelcast.jet.core.JobStatus.RUNNING;
@@ -95,28 +89,27 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
     @Parameters(name = "{2}")
     public static Collection<Object[]> data() {
-        return Arrays.asList(new Object[][] {
-                { RetryStrategies.never(), false, "fail"},
-                { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), false, "reconnect"},
-                { RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), true, "reconnect w/ state reset"}
+        return Arrays.asList(new Object[][]{
+                {RetryStrategies.never(), false, "fail"},
+                {RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), false, "reconnect"},
+                {RetryStrategies.indefinitely(RECONNECT_INTERVAL_MS), true, "reconnect w/ state reset"}
         });
     }
 
     @After
     public void after() {
         if (postgres != null) {
-            stopContainer(postgres);
+            postgres.stop();
         }
     }
 
     @Test
     public void when_noDatabaseToConnectTo() throws Exception {
-        postgres = initPostgres(null, 0);
+        postgres = initPostgres(null, null);
+        int port = fixPortBinding(postgres, POSTGRESQL_PORT);
         String containerIpAddress = postgres.getContainerIpAddress();
         stopContainer(postgres);
-        postgres = null;
 
-        int port = findRandomOpenPortInRange(POSTGRESQL_PORT + 100, POSTGRESQL_PORT + 1000);
         Pipeline pipeline = initPipeline(containerIpAddress, port);
 
         // when job starts
@@ -136,7 +129,7 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
             assertTrue(jet.getMap("results").isEmpty());
 
             // and DB starts
-            postgres = initPostgres(null, port);
+            startContainer(postgres);
             try {
                 // then source connects successfully
                 assertEqualsEventually(() -> jet.getMap("results").size(), 4);
@@ -186,8 +179,9 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
     @Test
     public void when_databaseShutdownOrLongDisconnectDuringSnapshotting() throws Exception {
-        int port = findRandomOpenPort();
-        postgres = initPostgres(null, port);
+        postgres = initPostgres(null, null);
+        int port = fixPortBinding(postgres, POSTGRESQL_PORT);
+
         Pipeline pipeline = initPipeline(postgres.getContainerIpAddress(), port);
         // when job starts
         JetInstance jet = createJetMembers(2)[0];
@@ -200,7 +194,6 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
         // and DB is stopped
         stopContainer(postgres);
-        postgres = null;
 
         // then
         boolean neverReconnect = reconnectBehavior.getMaxAttempts() == 0;
@@ -267,8 +260,9 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
     public void when_databaseShutdownOrLongDisconnectDuringBinlogReading() throws Exception {
         Assume.assumeFalse(reconnectBehavior.getMaxAttempts() < 0 && !resetStateOnReconnect);
 
-        int port = findRandomOpenPort();
-        postgres = initPostgres(null, port);
+        postgres = initPostgres(null, null);
+        int port = fixPortBinding(postgres, POSTGRESQL_PORT);
+
         Pipeline pipeline = initPipeline(postgres.getContainerIpAddress(), port);
         // when connector is up and transitions to binlog reading
         JetInstance jet = createJetMembers(2)[0];
@@ -280,7 +274,6 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
 
         // and DB is stopped
         stopContainer(postgres);
-        postgres = null;
 
         boolean neverReconnect = reconnectBehavior.getMaxAttempts() == 0;
         if (neverReconnect) {
@@ -358,8 +351,7 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
         if (network != null) {
             postgres = postgres.withNetwork(network);
         }
-        postgres.start();
-        waitUntilUp(postgres);
+        startContainer(postgres);
         return postgres;
     }
 
@@ -378,24 +370,6 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
         return toxiproxy.getProxy(postgres, POSTGRESQL_PORT);
     }
 
-    private static void waitUntilUp(PostgreSQLContainer<?> postgres) {
-        for (; ; ) {
-            try {
-                try (Connection connection = DriverManager.getConnection(postgres.getJdbcUrl(), postgres.getUsername(),
-                        postgres.getPassword())) {
-                    connection.setSchema("inventory");
-                    boolean successfull = connection.createStatement().execute("SELECT 1");
-                    if (successfull) {
-                        return;
-                    }
-                }
-            } catch (SQLException throwables) {
-                // repeat
-            }
-            System.out.println("Waiting for the database to come up...");
-        }
-    }
-
     private static void insertRecords(PostgreSQLContainer<?> postgres, int... ids) throws SQLException {
         try (Connection connection = DriverManager.getConnection(postgres.getJdbcUrl(), postgres.getUsername(),
                 postgres.getPassword())) {
@@ -409,31 +383,6 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
             statement.executeBatch();
             connection.commit();
         }
-    }
-
-    private static int findRandomOpenPort() throws Exception {
-        try (ServerSocket socket = new ServerSocket(0)) {
-            return socket.getLocalPort();
-        }
-    }
-
-    private static int findRandomOpenPortInRange(int fromInclusive, int toExclusive) throws IOException {
-        List<Integer> randomizedPortsInRange = IntStream.range(fromInclusive, toExclusive)
-                .boxed()
-                .collect(Collectors.toList());
-        Collections.shuffle(randomizedPortsInRange);
-
-        for (int port : randomizedPortsInRange) {
-            try {
-                ServerSocket serverSocket = new ServerSocket(port);
-                serverSocket.close();
-                return port;
-            } catch (Exception e) {
-                //swallow, expected
-            }
-        }
-
-        throw new IOException("No free port in range [" + fromInclusive + ", " + toExclusive + ")");
     }
 
 }

--- a/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
+++ b/extensions/cdc-postgres/src/test/java/com/hazelcast/jet/cdc/postgres/PostgresCdcNetworkIntegrationTest.java
@@ -129,7 +129,7 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
             assertTrue(jet.getMap("results").isEmpty());
 
             // and DB starts
-            startContainer(postgres);
+            postgres.start();
             try {
                 // then source connects successfully
                 assertEqualsEventually(() -> jet.getMap("results").size(), 4);
@@ -351,7 +351,7 @@ public class PostgresCdcNetworkIntegrationTest extends AbstractCdcIntegrationTes
         if (network != null) {
             postgres = postgres.withNetwork(network);
         }
-        startContainer(postgres);
+        postgres.start();
         return postgres;
     }
 

